src/main/scala/firrtl/transforms/ConstantPropagation.scala

```diff
@@ -481,17 +481,23 @@ class ConstantPropagation extends Transform with ResolvedAnnotationPaths {
       propagated
     }
 
-    def backPropStmt(stmt: Statement): Statement = stmt map backPropExpr match {
-      case decl: IsDeclaration if swapMap.contains(decl.name) =>
-        val newName = swapMap(decl.name)
-        nPropagated += 1
-        decl match {
-          case node: DefNode => node.copy(name = newName)
-          case wire: DefWire => wire.copy(name = newName)
-          case reg: DefRegister => reg.copy(name = newName)
-          case other => throwInternalError()
-        }
-      case other => other map backPropStmt
+    def backPropStmt(stmt: Statement): Statement = stmt match {
+      case reg: DefRegister if (WrappedExpression.weq(reg.init, WRef(reg))) =>
+        // Self-init reset is an idiom for "no reset," and must be handled separately
+        val updatedName = swapMap.getOrElse(reg.name, reg.name)
+        reg.copy(name = updatedName, init = WRef(reg).copy(name = updatedName))
```

Comment:
```suggestion
        swapMap.get(reg.name)
               .map(newName => reg.copy(name = newName, init = WRef(reg).copy(name = newName)))
               .getOrElse(reg)
```
I'm not sure if it's worth trying to optimize these object allocations since every `map whatever` is returning a new object even if it doesn't change anything, but we could.

Reply:
Whenever I "peephole fix" code I try to more or less make new cases blindly resemble the style of the existing cases, but this is your code and I am generally a fan of relying on `None.map` for the common case of non-const-propped registers :)